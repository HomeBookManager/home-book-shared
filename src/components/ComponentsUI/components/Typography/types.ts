// @home-book
import {
  COMMON_COLORS,
  DARK_COLORS,
  FORM_COLORS,
  LIGHT_COLORS,
} from 'home-book-constants';
import { TValueObj, Theme } from 'home-book-types';

// types
import { TTypographyProps } from './Typography';

export type TTypographyColor = TValueObj<
  typeof COMMON_COLORS & typeof FORM_COLORS
>;

export type TTypographyColorMode = {
  [Theme.dark]: TValueObj<typeof DARK_COLORS>;
  [Theme.light]: TValueObj<typeof LIGHT_COLORS>;
};

export type TOmittedTypographyProps = Omit<TTypographyProps, 'fontType'>;
