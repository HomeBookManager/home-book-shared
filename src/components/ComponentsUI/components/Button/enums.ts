export enum ButtonColor {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  warning = 'warning',
  error = 'error',
}

export enum ButtonSize {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

export enum ButtonVariant {
  text = 'text',
  contained = 'contained',
  outlined = 'outlined',
}
