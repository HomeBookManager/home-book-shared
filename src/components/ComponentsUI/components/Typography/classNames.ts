import { mapValues } from 'lodash';

// types
import {
  TypographyFontStyle,
  TypographyFontWeight,
  TypographyFontType,
} from './enums';

export const className = 'Typography';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      noWrap: `${className}--no-wrap`,
      withoutMargin: `${className}--without-margin`,
      ...mapValues(
        TypographyFontType,
        (fontType) => `${className}--${fontType}`,
      ),
      ...mapValues(
        TypographyFontStyle,
        (fontStyle) => `${className}--${fontStyle}`,
      ),
      ...mapValues(
        TypographyFontWeight,
        (fontWeight) => `${className}--${fontWeight}`,
      ),
    },
  },
};
