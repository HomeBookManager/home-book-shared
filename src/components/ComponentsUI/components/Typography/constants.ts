// @home-book
import { DARK_COLORS, LIGHT_COLORS } from 'home-book-constants';
import { Theme } from 'home-book-types';

export const DEFAULT_COLOR_MODE = {
  [Theme.dark]: DARK_COLORS.neutral1,
  [Theme.light]: LIGHT_COLORS.neutral1,
};
