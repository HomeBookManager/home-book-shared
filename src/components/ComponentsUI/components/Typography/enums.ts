export enum TypographyFontStyle {
  italic = 'italic',
  normal = 'normal',
}

export enum TypographyFontType {
  h1 = 'h1',
  h2 = 'h2',
  h3 = 'h3',
  h4 = 'h4',
  h5 = 'h5',
  h6 = 'h6',
  p = 'p',
  small = 'small',
}

export enum TypographyFontWeight {
  black = 'black',
  bold = 'bold',
  extraBold = 'extra-bold',
  light = 'light',
  medium = 'medium',
  regular = 'regular',
  semiBold = 'semi-bold',
}
