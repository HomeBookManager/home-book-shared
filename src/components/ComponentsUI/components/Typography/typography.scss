@use 'sass:map';

@import '../../../../styles/mixins/text-ellipsis';
@import '../../../../styles/mixins/typography/typography';
@import '../../../../styles/mixins/typography/typography-font-family';
@import '../../../../styles/mixins/typography/typography-list';
@import '../../../../styles/mixins/typography/typography-map';

.Typography {
  @each $font-type in $--cth-font-types {
    &--#{$font-type} {
      $--cth-font-size: map.get($--cth-font-size-map, $font-type);
      $--cth-line-height: map.get($--cth-line-height-map, $font-type);
      $--cth-margin-bottom: map.get($--cth-margin-bottom-map, $font-type);

      @if $font-type == 'p' or $font-type == 'small' {
        @include standard;
      } @else {
        @include heading;
      }

      @include typography-styles(
        $--cth-font-size,
        $--cth-line-height,
        $--cth-margin-bottom
      );
    }
  }

  @each $font-style in $--cth-font-styles {
    &--#{$font-style} {
      font-style: map.get($--cth-font-styles-map, $font-style);
    }
  }

  @each $font-weight in $--cth-font-weights {
    &--#{$font-weight} {
      /* stylelint-disable-next-line font-weight-notation */
      font-weight: map.get($--cth-font-weights-map, $font-weight);
    }
  }

  &--no-wrap {
    @include text-ellipsis;
  }

  &--without-margin {
    margin: 0;
  }
}
