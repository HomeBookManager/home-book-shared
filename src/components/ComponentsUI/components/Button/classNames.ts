import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';

export const className = 'Button';

export const getIconModificators = (property: 'fill' | 'stroke') => ({
  ...mapValues(ButtonColor, (color) => ({
    ...mapValues(
      ButtonVariant,
      (variant) =>
        `${className}__icon--apply-${property}--${variant}--${color}`,
    ),
  })),
});

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      fullwidth: `${className}--full-width`,
      forcedHover: {
        ...mapValues(ButtonVariant, (variant) => ({
          ...mapValues(
            ButtonColor,
            (color) => `${className}--${variant}--${color}--forced-hover`,
          ),
        })),
      },
      ...mapValues(ButtonColor, (color) => `${className}--${color}`),
      ...mapValues(ButtonSize, (size) => `${className}--${size}`),
      ...mapValues(ButtonVariant, (variant) => `${className}--${variant}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      applyFill: `${className}__icon--apply-fill`,
      applyStroke: `${className}__icon--apply-stroke`,
      end: `${className}__icon--end`,
      start: `${className}__icon--start`,
      ...mapValues(ButtonSize, (size) => `${className}__icon--${size}`),
    },
  },
};
