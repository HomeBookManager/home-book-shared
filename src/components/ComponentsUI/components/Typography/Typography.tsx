import classnames from 'classnames';
import { camelCase } from 'lodash';
import { createElement, FC, HTMLAttributes, ReactNode } from 'react';

// hooks
import { useTheme } from '../../../../hooks';

// others
import { className as classNameTypography, classNames } from './classNames';
import { DEFAULT_COLOR_MODE } from './constants';

// styles
import './typography.scss';

// types
import { TTypographyColor, TTypographyColorMode } from './types';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import {
  TypographyFontStyle,
  TypographyFontWeight,
  TypographyFontType,
} from './enums';

// utils
import { getDataTestAttribute } from '../../../E2EDataAttributes/utils';

export type TTypographyProps = {
  children?: ReactNode;
  color?: TTypographyColor;
  colorMode?: TTypographyColorMode;
  e2eAttribute?: string;
  e2eValue?: number | string;
  fontStyle?: TypographyFontStyle;
  fontType?: TypographyFontType;
  fontWeight?: TypographyFontWeight;
  innerHtml?: string;
  withoutMargin?: boolean;
} & HTMLAttributes<HTMLElement>;

export const Typography: FC<TTypographyProps> = ({
  children,
  className = '',
  color = '',
  colorMode = DEFAULT_COLOR_MODE,
  e2eAttribute = E2EAttribute.text,
  e2eValue = '',
  fontStyle = TypographyFontStyle.normal,
  fontType = TypographyFontType.p,
  fontWeight = TypographyFontWeight.regular,
  innerHtml = '',
  withoutMargin = true,
  style = {},
  ...restProps
}) => {
  const { classNamesWithTheme, theme } = useTheme(classNames);

  if (!children && !innerHtml) {
    return null;
  }

  return createElement(
    fontType,
    {
      ...restProps,
      ...(innerHtml ? { dangerouslySetInnerHTML: { __html: innerHtml } } : {}),
      [getDataTestAttribute(e2eAttribute)]: e2eValue,
      className: classnames(
        classNamesWithTheme[classNameTypography].name,
        className,
        classNamesWithTheme[classNameTypography].modificators[
          camelCase(fontWeight) as keyof typeof TypographyFontWeight
        ],
        classNamesWithTheme[classNameTypography].modificators[fontStyle],
        classNamesWithTheme[classNameTypography].modificators[fontType],
        {
          [classNamesWithTheme[classNameTypography].modificators.withoutMargin]:
            withoutMargin,
        },
      ),
      style: {
        ...style,
        color: color || colorMode[theme],
      },
    },
    children,
  );
};

export default Typography;
