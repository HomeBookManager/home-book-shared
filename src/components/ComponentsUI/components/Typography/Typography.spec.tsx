import { render } from '@testing-library/react';

// @home-book
import { COMMON_COLORS, LIGHT_COLORS } from 'home-book-constants';

// components
import Typography from './Typography';

// others
import { className as classNameTypography, classNames } from './classNames';
import { TYPOGRAPHY_COLORS_MODE } from './constants';

// types
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import {
  TypographyFontStyle,
  TypographyFontType,
  TypographyFontWeight,
} from './enums';

// utils
import { getByE2EAttribute } from '../../../../tests/testHelpers';
import { getDataTestAttribute } from '../../../E2EDataAttributes/utils';
import { hexToRgb } from '../../../../utils/transform/hexToRgb/hexToRgb';

const className = 'className';

describe('Typography props', () => {
  it('should pass align', () => {
    // before
    const align = 'center';
    const rgb = hexToRgb(LIGHT_COLORS.neutral1);

    const { container } = render(
      <Typography align={align}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      'style',
      `color: rgb(${rgb.r}, ${rgb.g}, ${rgb.b}); text-align: ${align};`,
    );
  });

  it('should pass children', () => {
    // mock
    const children = 'children';

    //before
    const { container } = render(<Typography>{children}</Typography>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveTextContent(
      children,
    );
  });

  it('should pass className', () => {
    //before
    const { container } = render(
      <Typography className={className}>children</Typography>,
    );

    // result
    expect(container.querySelector(`.${className}`)).toHaveClass(className);
  });

  it('should pass color', () => {
    //before
    const rgb1 = hexToRgb(COMMON_COLORS.blue1);
    const rgb2 = hexToRgb(LIGHT_COLORS.neutral5);

    const { container: container1 } = render(
      <Typography color={COMMON_COLORS.blue1}>children</Typography>,
    );

    const { container: container2 } = render(
      <Typography color={TYPOGRAPHY_COLORS_MODE.neutral5}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container1, E2EAttribute.text)).toHaveAttribute(
      'style',
      `color: rgb(${rgb1.r}, ${rgb1.g}, ${rgb1.b}); text-align: inherit;`,
    );

    expect(getByE2EAttribute(container2, E2EAttribute.text)).toHaveAttribute(
      'style',
      `color: rgb(${rgb2.r}, ${rgb2.g}, ${rgb2.b}); text-align: inherit;`,
    );
  });

  it('should pass e2eAttribute', () => {
    // before
    const { container } = render(
      <Typography e2eAttribute={E2EAttribute.text}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.text),
    );
  });

  it('should pass e2eValue', () => {
    // mock
    const e2eValue = 'e2eValue';

    // before
    const { container } = render(
      <Typography e2eValue={e2eValue}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.text),
      e2eValue,
    );
  });

  it('should pass fontStyle', () => {
    //before
    const { container } = render(
      <Typography fontStyle={TypographyFontStyle.italic}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveClass(
      classNames[classNameTypography].modificators[TypographyFontStyle.italic],
    );
  });

  it('should pass fontType', () => {
    //before
    const { container } = render(
      <Typography fontType={TypographyFontType.h1}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveClass(
      classNames[classNameTypography].modificators[TypographyFontType.h1],
    );

    expect(
      container.getElementsByTagName(TypographyFontType.h1)[0],
    ).toBeDefined();
  });

  it('should pass fontWeight', () => {
    //before
    const { container } = render(
      <Typography fontWeight={TypographyFontWeight.black}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveClass(
      classNames[classNameTypography].modificators[TypographyFontWeight.black],
    );
  });

  it('should pass innerHtml', () => {
    // mock
    const children = 'children';
    const htmlTag = 'span';

    //before
    const { container } = render(
      <Typography innerHtml={`<${htmlTag}>${children}</${htmlTag}>`} />,
    );

    // result
    expect(container.getElementsByTagName(htmlTag)[0]).toBeDefined();

    expect(container.getElementsByTagName(htmlTag)[0]).toHaveTextContent(
      children,
    );
  });

  it('should pass noWrap', () => {
    //before
    const { container } = render(<Typography noWrap>children</Typography>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveClass(
      classNames[classNameTypography].modificators.noWrap,
    );
  });

  it('should pass style', () => {
    // before
    const rgb = hexToRgb(LIGHT_COLORS.neutral1);

    const { container } = render(
      <Typography style={{ width: '100%' }}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      'style',
      `width: 100%; color: rgb(${rgb.r}, ${rgb.g}, ${rgb.b}); text-align: inherit;`,
    );
  });

  it('should pass withoutMargin', () => {
    //before
    const { container } = render(
      <Typography withoutMargin={false}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).not.toHaveClass(
      classNames[classNameTypography].modificators.withoutMargin,
    );
  });
});

describe('Typography snapshots', () => {
  it('should render empty', () => {
    //before
    const { asFragment } = render(<Typography />);

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with some text', () => {
    // before
    const { asFragment } = render(<Typography>Text</Typography>);

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with passed inner html', () => {
    // before
    const { asFragment } = render(
      <Typography innerHtml={'<div>innerHtml</div>'} />,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
