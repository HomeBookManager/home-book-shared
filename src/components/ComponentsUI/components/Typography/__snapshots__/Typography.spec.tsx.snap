// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography snapshots should render empty 1`] = `<DocumentFragment />`;

exports[`Typography snapshots should render with passed inner html 1`] = `
<DocumentFragment>
  <p
    class="Typography Typography--regular Typography--normal Typography--p Typography--without-margin"
    data-test-text=""
    style="color: rgb(30, 40, 57); text-align: inherit;"
  />
  <div>
    innerHtml
  </div>
  <p />
</DocumentFragment>
`;

exports[`Typography snapshots should render with some text 1`] = `
<DocumentFragment>
  <p
    class="Typography Typography--regular Typography--normal Typography--p Typography--without-margin"
    data-test-text=""
    style="color: rgb(30, 40, 57); text-align: inherit;"
  >
    Text
  </p>
</DocumentFragment>
`;
